Add pipeline framework

jenniferjiangkells opened the pull request:
# Description
Addresses #54

Introduces the concept of `Pipeline`, `Component`, and `DataContainers`.

These are the building blocks of the pipelineing component of HC. We can give the users 3 levels of control:
1. Build your own pipeline, using inline functions - this is the easiest and most flexible level, good for quick experiments
2. Build your own pipeline, using `Component` classes - this adds an extra layer of abstraction, especially useful for wrapping specific models such as MedCAT, ClinicalBERT, LLMs.
3. Use prebuilt pipelines e.g. `MedicalCodingPipeline` - prebuilt pipelines are a pre-configured set of components for specific use cases and has the highest level of abstraction. This is the easiest to get up and running with something functional.

Extra - loading pipeline integrations from other libraries such as spacy, huggingface, etc. #27 

Went over-scope a little and additionally implemented `TextPreprocessor`, `Model`, `TextPostprocessor`, and `MedicalCodingPipeline`, which are implementations of `Component` and `Pipeline`. Helps to see what I want the downstream usage to look like.

It's probably best to introduce new concepts through examples, so here's a code snippet:

```python
from healthchain.io.containers import Document
from healthchain.pipeline import Pipeline
from healthchain.pipeline.components import Model
from healthchain.pipeline.components import TextPostProcessor
from healthchain.pipeline.components import TextPreprocessor
from healthchain.pipeline import MedicalCodingPipeline

################################################################################
# 1. Build your own pipeline, using inline functions
################################################################################

# initialise the pipeline with the data type you want to process
nlp_pipeline = Pipeline[Document]()

@nlp_pipeline.add(stage="preprocessing")
def tokenize(doc: Document) -> Document:
    doc.tokens = doc.text.split()
    return doc


@nlp_pipeline.add(stage="preprocessing", dependencies=["tokenize"])
def pos_tag(doc: Document) -> Document:
    # Dummy POS tagging
    doc.pos_tags = ["NOUN" if token[0].isupper() else "VERB" for token in doc.tokens]
    return doc


@nlp_pipeline.add(dependencies=["tokenize", "pos_tag"])
def ner(doc: Document) -> Document:
    # Dummy NER
    doc.entities = [
        token for token, pos in zip(doc.tokens, doc.pos_tags) if pos == "NOUN"
    ]
    return doc

print("Initial pipeline:")
print(nlp_pipeline)
print(nlp_pipeline.stages)

@nlp_pipeline.add(position="after", reference="tokenize")
def remove_stopwords(doc: Document) -> Document:
    stopwords = {"the", "a", "an", "in", "on", "at"}
    doc.tokens = [token for token in doc.tokens if token not in stopwords]
    return doc

print("After adding remove_stopwords:")
print(nlp_pipeline)
print(nlp_pipeline.stages)


# Remove method
def new_tokenizer(doc: Document) -> Document:
    doc.tokens = doc.text.split() + ["<EOS>"]  # Add end-of-sentence token
    return doc


nlp_pipeline.remove("tokenize")
nlp_pipeline.add(new_tokenizer, name="tokenize", position="first")

# Replace method
def advanced_ner(doc: Document) -> Document:
    # More sophisticated NER logic
    doc.entities = [
        token for token in doc.tokens if token[0].isupper() and len(token) > 1
    ]
    return doc

nlp_pipeline.replace("ner", advanced_ner)

print("After replacing ner:")
print(nlp_pipeline)
print(nlp_pipeline.stages)

# Usage
# NLP pipeline
nlp = nlp_pipeline.build()

doc = Document("OpenAI released GPT-4 in 2023.")

result = nlp(doc)
print(f"Char count: {doc.char_count()}")
print(f"Word count: {doc.word_count()}")
print(f"Tokens: {result.tokens}")
print(f"POS Tags: {result.pos_tags}")
print(f"NER: {result.get_entities()}")

preprocessing_components = nlp_pipeline._stages.get("preprocessing", [])
print(f"Preprocessing components: {[c.__name__ for c in preprocessing_components]}")


################################################################################
# 2. Build your own pipeline, using Component classes (or mix and match)
################################################################################
component_pipeline = Pipeline[Document]()

component_pipeline.add(TextPreprocessor())
component_pipeline.add(Model(model_path="path/to/model"))
component_pipeline.add(TextPostProcessor())
component_pipeline.add(remove_stopwords, position="last")

# Or this is how you would configure it, not sure about adding an extra config, seems a bit clunky, might remove
# postprocessor_config = TextPostProcessorConfig(
#     postcoordination_lookup={
#         "heart attack": "myocardial infarction",
#         "high blood pressure": "hypertension"
#     }
# )
# component_pipeline.add(TextPostProcessor(postprocessor_config))

components = component_pipeline.build()
result = components(doc)

print(component_pipeline)
print(component_pipeline.stages)
print(f"Tokens: {result.tokens}")
print(f"POS Tags: {result.pos_tags}")
print(f"NER: {result.entities}")


################################################################################
# 3. Use prebuilt pipelines e.g. MedicalCodingPipeline
################################################################################
pipeline = MedicalCodingPipeline.load("./path/to/model")

coding_pipeline = pipeline.build()
result = coding_pipeline(doc)

print(pipeline)
print(pipeline.stages)
print(f"Processed Text: {result.text}")
print(f"Tokens: {result.tokens}")
print(f"Entities: {result.entities}")
```

jenniferjiangkells commented:
@adamkells dw im going to explain everything ☝️

adamkells commented on healthchain/pipeline/__init__.py:
```diff
@@ -0,0 +1,17 @@
+    "BaseComponent",
+    "Component",
+    "Model",
+    "TextPreprocessor",
```
Should be TextPreProcessor for consistency with casing of post processor.

adamkells commented on healthchain/pipeline/basepipeline.py:
```diff
@@ -0,0 +1,463 @@
+        if isinstance(component, BaseComponent):
+            return wrapper(component)
+        elif isinstance(component, Component):
+            return wrapper(component.__call__)
```
I don't understand why component needs __call__ if it's not a basecomponent.

jenniferjiangkells replied:
good point

adamkells commented on healthchain/pipeline/basepipeline.py:
```diff
@@ -0,0 +1,463 @@
+        """
+        # Check if the component exists in the pipeline
+        if not any(c.name == component_name for c in self._components):
+            logger.warning(f"Component '{component_name}' not found in the pipeline.")
```
No real issue here but I'm just wondering whether this edge case should fail louder? 

Specifically I see this warning being triggered by typos on r some underlying mistake. I think the user in that case would want to be explicitly thrown an error to help them catch the mistake.

jenniferjiangkells replied:
good point

adamkells commented on healthchain/pipeline/basepipeline.py:
```diff
@@ -0,0 +1,463 @@
+        if ref_index is None:
+            raise ValueError(f"Reference component '{reference}' not found.")
+
+        self._components.insert(ref_index + offset, component)
```
I could have missed it but what happens if I add a stage with the name of a stage which exists already? Can i have duplicates? Or should stage names be forced to be unique?

jenniferjiangkells replied:
it would just be appended to the existing stage, no?

adamkells commented on healthchain/pipeline/basepipeline.py:
```diff
@@ -0,0 +1,463 @@
+
+        if not old_component_found:
+            logger.warning(
+                f"Component '{old_component_name}' not found in the pipeline."
```
Again I wonder if not failing here might increase the likelihood of users get unintended behaviour.

jenniferjiangkells replied:
gp

adamkells commented on healthchain/pipeline/basepipeline.py:
```diff
@@ -0,0 +1,463 @@
+                        resolved.append(component)
+                        unresolved.remove(component)
+                        break
+                else:
```
Could be my phone but indentation of the else looks off to me.

jenniferjiangkells replied:
it is your phone

adamkells commented on healthchain/pipeline/basepipeline.py:
```diff
@@ -0,0 +1,463 @@
+        def pipeline(data: Union[T, DataContainer[T]]) -> DataContainer[T]:
+            if not isinstance(data, DataContainer):
+                data = DataContainer(data)
+            return reduce(lambda d, comp: comp(d), ordered_components, data)
```
I don't understand what is happening on this line, can you explain?

jenniferjiangkells replied:
it's basically doing this but more concisely with `reduce`:
```python
for component in ordered_components:
    result = component(result)
return result
```
it also checks that the input is an instance of `DataContainer` or its subclass. if not, it wraps it in a generic container. wrapping the input in a container doesn't do much now but hopefully will expand functionality.

adamkells reviewed (commented):
I think this is great. There's a fair bit of complexity but it's all hidden away in such a way that the user doesn't need to worry. This is really impressive and a great next step for the project. I've left a few minor comments but overall great job!

jenniferjiangkells commented:
@adamkells addressed all your comments, don't think you need to go over the code but feel free to check the documentation as I made quite a lot of changes there

jenniferjiangkells commented:
@adamkells my man can i just merge this

adamkells commented:
Oh go on then
